File: packages/dxtr_pixs_gpu/android/src/main/kotlin/dev/pixelcraft/pixelcraft/CameraLookLutComposer.kt
package dev.pixelcraft.pixelcraft

import android.content.Context
import kotlin.math.floor
import kotlin.math.roundToInt

/**
 * PF2 control-rate LUT composer for the Android live camera path.
 *
 * Composition is evaluated only when CameraLook state changes. The 60 fps
 * renderer continues sampling one 33^3 LUT atlas per frame, preserving the
 * proven OES hot path. Saved pixels remain Rust-authoritative.
 *
 * Canonical order:
 *   Adjust -> Film -> Creative
 */
internal class CameraLookLutComposer(context: Context) {
    companion object {
        const val LUT_SIZE = 33
        const val TILES = 6
        const val ATLAS_SIZE = LUT_SIZE * TILES
        private const val MIDPOINT = 128f / 255f
    }

    private val appContext = context.applicationContext

    fun compose(look: NativeGpuCameraLook): ByteArray {
        val film = if (look.hasFilm) loadAtlas(look.filmProfileId) else null
        val creativeAsset = NativeGpuCameraLook.creativeLutAssetId(look.creativeFilterId)
        val creative = if (look.hasCreative && creativeAsset != null) loadAtlas(creativeAsset) else null
        val output = ByteArray(ATLAS_SIZE * ATLAS_SIZE * 4)

        for (blue in 0 until LUT_SIZE) {
            for (green in 0 until LUT_SIZE) {
                for (red in 0 until LUT_SIZE) {
                    var color = floatArrayOf(
                        red.toFloat() / (LUT_SIZE - 1),
                        green.toFloat() / (LUT_SIZE - 1),
                        blue.toFloat() / (LUT_SIZE - 1),
                    )

                    color = applyAdjustments(color, look)

                    if (film != null) {
                        color = mix(color, sampleAtlas(film, color), look.filmStrength)
                    }

                    if (look.hasCreative) {
                        color = when (look.creativeFilterId) {
                            "grayscale", "invert" -> applyExactCreative(
                                color,
                                look.creativeFilterId,
                                look.creativeFilterStrength,
                            )
                            else -> {
                                check(creative != null) {
                                    "Missing canonical creative LUT for ${look.creativeFilterId}"
                                }
                                mix(color, sampleAtlas(creative, color), look.creativeFilterStrength)
                            }
                        }
                    }

                    writeAtlasTexel(output, red, green, blue, color)
                }
            }
        }
        return output
    }

    private fun applyAdjustments(
        source: FloatArray,
        look: NativeGpuCameraLook,
    ): FloatArray {
        var r = clamp01(source[0] + (look.brightness - 1f))
        var g = clamp01(source[1] + (look.brightness - 1f))
        var b = clamp01(source[2] + (look.brightness - 1f))

        r = clamp01((r - MIDPOINT) * look.contrast + MIDPOINT)
        g = clamp01((g - MIDPOINT) * look.contrast + MIDPOINT)
        b = clamp01((b - MIDPOINT) * look.contrast + MIDPOINT)

        val luminance = r * 0.2126f + g * 0.7152f + b * 0.0722f
        r = clamp01(luminance + (r - luminance) * look.saturation)
        g = clamp01(luminance + (g - luminance) * look.saturation)
        b = clamp01(luminance + (b - luminance) * look.saturation)
        return floatArrayOf(r, g, b)
    }

    /** Mirrors the existing Metal editor parity path: rounded u8 math. */
    private fun applyExactCreative(
        source: FloatArray,
        filterId: String,
        strength: Float,
    ): FloatArray {
        val source8 = IntArray(3) { channel ->
            (clamp01(source[channel]) * 255f).roundToInt().coerceIn(0, 255)
        }
        val effected8 = when (filterId) {
            "grayscale" -> {
                val average = (source8[0] + source8[1] + source8[2]) / 3
                intArrayOf(average, average, average)
            }
            "invert" -> intArrayOf(255 - source8[0], 255 - source8[1], 255 - source8[2])
            else -> error("Unsupported exact creative filter: $filterId")
        }
        val amount = strength.coerceIn(0f, 1f)
        return FloatArray(3) { channel ->
            val blended = (
                source8[channel] +
                    (effected8[channel] - source8[channel]) * amount
                ).roundToInt().coerceIn(0, 255)
            blended / 255f
        }
    }

    private fun loadAtlas(assetId: String): ByteArray =
        appContext.assets.open("gpu_luts/$assetId.rgba8").use { stream ->
            stream.readBytes().also { bytes ->
                check(bytes.size == ATLAS_SIZE * ATLAS_SIZE * 4) {
                    "Unexpected LUT atlas size for $assetId: ${bytes.size}"
                }
            }
        }

    private fun sampleAtlas(atlas: ByteArray, color: FloatArray): FloatArray {
        val r = clamp01(color[0]) * (LUT_SIZE - 1)
        val g = clamp01(color[1]) * (LUT_SIZE - 1)
        val b = clamp01(color[2]) * (LUT_SIZE - 1)
        val r0 = floor(r).toInt()
        val g0 = floor(g).toInt()
        val b0 = floor(b).toInt()
        val r1 = (r0 + 1).coerceAtMost(LUT_SIZE - 1)
        val g1 = (g0 + 1).coerceAtMost(LUT_SIZE - 1)
        val b1 = (b0 + 1).coerceAtMost(LUT_SIZE - 1)
        val rf = r - r0
        val gf = g - g0
        val bf = b - b0

        fun slice(blue: Int): FloatArray {
            val c00 = texel(atlas, r0, g0, blue)
            val c10 = texel(atlas, r1, g0, blue)
            val c01 = texel(atlas, r0, g1, blue)
            val c11 = texel(atlas, r1, g1, blue)
            return FloatArray(3) { channel ->
                val top = c00[channel] + (c10[channel] - c00[channel]) * rf
                val bottom = c01[channel] + (c11[channel] - c01[channel]) * rf
                top + (bottom - top) * gf
            }
        }

        val low = slice(b0)
        val high = slice(b1)
        return FloatArray(3) { channel -> low[channel] + (high[channel] - low[channel]) * bf }
    }

    private fun texel(atlas: ByteArray, red: Int, green: Int, blue: Int): FloatArray {
        val tileX = blue % TILES
        val tileY = blue / TILES
        val x = tileX * LUT_SIZE + red
        val y = tileY * LUT_SIZE + green
        val offset = (y * ATLAS_SIZE + x) * 4
        return floatArrayOf(
            (atlas[offset].toInt() and 0xff) / 255f,
            (atlas[offset + 1].toInt() and 0xff) / 255f,
            (atlas[offset + 2].toInt() and 0xff) / 255f,
        )
    }

    private fun writeAtlasTexel(
        atlas: ByteArray,
        red: Int,
        green: Int,
        blue: Int,
        color: FloatArray,
    ) {
        val tileX = blue % TILES
        val tileY = blue / TILES
        val x = tileX * LUT_SIZE + red
        val y = tileY * LUT_SIZE + green
        val offset = (y * ATLAS_SIZE + x) * 4
        atlas[offset] = channelByte(color[0])
        atlas[offset + 1] = channelByte(color[1])
        atlas[offset + 2] = channelByte(color[2])
        atlas[offset + 3] = 0xff.toByte()
    }

    private fun mix(source: FloatArray, effected: FloatArray, amount: Float): FloatArray {
        val t = amount.coerceIn(0f, 1f)
        return FloatArray(3) { channel ->
            clamp01(source[channel] + (effected[channel] - source[channel]) * t)
        }
    }

    private fun channelByte(value: Float): Byte =
        (clamp01(value) * 255f).roundToInt().coerceIn(0, 255).toByte()

    private fun clamp01(value: Float): Float = value.coerceIn(0f, 1f)
}

File: packages/dxtr_pixs_gpu/ios/Classes/CameraLookLutComposer.swift
import Foundation
import Metal

/// PF2 control-rate LUT composer for the iOS live camera path.
///
/// Composition happens only when CameraLook changes. The 60 fps Metal renderer
/// continues sampling a single 33^3 LUT texture per frame. Saved pixels remain
/// Rust-authoritative.
///
/// Canonical order: Adjust -> Film -> Creative.
final class CameraLookLutComposer {
  private static let lutSize = 33
  private static let tiles = 6
  private static let atlasSize = lutSize * tiles
  private static let midpoint: Float = 128.0 / 255.0

  private let device: MTLDevice

  init(device: MTLDevice) {
    self.device = device
  }

  func compose(_ look: NativeGpuCameraLook) throws -> MTLTexture {
    let film = look.hasFilm ? try loadAtlas(look.filmProfileId) : nil
    let creativeAsset = look.creativeLutAssetId
    let creative = look.hasCreative && creativeAsset != nil
      ? try loadAtlas(creativeAsset!)
      : nil

    var volume = Data(count: Self.lutSize * Self.lutSize * Self.lutSize * 4)
    volume.withUnsafeMutableBytes { raw in
      guard let bytes = raw.bindMemory(to: UInt8.self).baseAddress else { return }
      for blue in 0..<Self.lutSize {
        for green in 0..<Self.lutSize {
          for red in 0..<Self.lutSize {
            var color = SIMD3<Float>(
              Float(red) / Float(Self.lutSize - 1),
              Float(green) / Float(Self.lutSize - 1),
              Float(blue) / Float(Self.lutSize - 1)
            )

            color = Self.applyAdjustments(color, look: look)

            if let film {
              let effected = Self.sampleAtlas(film, color: color)
              color = Self.mix(color, effected, amount: look.filmStrength)
            }

            if look.hasCreative {
              if look.creativeFilterId == "grayscale" || look.creativeFilterId == "invert" {
                color = Self.applyExactCreative(
                  color,
                  filterId: look.creativeFilterId,
                  strength: look.creativeFilterStrength
                )
              } else if let creative {
                let effected = Self.sampleAtlas(creative, color: color)
                color = Self.mix(color, effected, amount: look.creativeFilterStrength)
              }
            }

            let offset = (red + Self.lutSize * (green + Self.lutSize * blue)) * 4
            bytes[offset] = Self.channelByte(color.x)
            bytes[offset + 1] = Self.channelByte(color.y)
            bytes[offset + 2] = Self.channelByte(color.z)
            bytes[offset + 3] = 255
          }
        }
      }
    }

    return try makeTexture(bytes: volume)
  }

  private func loadAtlas(_ assetId: String) throws -> Data {
    guard let url = Bundle.main.url(
      forResource: assetId,
      withExtension: "rgba8",
      subdirectory: "gpu_luts"
    ) else {
      throw Self.error("Missing camera look LUT asset: \(assetId)")
    }
    let data = try Data(contentsOf: url)
    let expected = Self.atlasSize * Self.atlasSize * 4
    guard data.count == expected else {
      throw Self.error("Unexpected camera look LUT byte count for \(assetId): \(data.count)")
    }
    return data
  }

  private func makeTexture(bytes: Data) throws -> MTLTexture {
    let descriptor = MTLTextureDescriptor()
    descriptor.textureType = .type3D
    descriptor.pixelFormat = .rgba8Unorm
    descriptor.width = Self.lutSize
    descriptor.height = Self.lutSize
    descriptor.depth = Self.lutSize
    descriptor.mipmapLevelCount = 1
    descriptor.usage = [.shaderRead]
    descriptor.storageMode = .shared

    guard let texture = device.makeTexture(descriptor: descriptor) else {
      throw Self.error("Unable to allocate composed 33^3 camera look LUT")
    }

    bytes.withUnsafeBytes { raw in
      guard let base = raw.baseAddress else { return }
      texture.replace(
        region: MTLRegionMake3D(0, 0, 0, Self.lutSize, Self.lutSize, Self.lutSize),
        mipmapLevel: 0,
        slice: 0,
        withBytes: base,
        bytesPerRow: Self.lutSize * 4,
        bytesPerImage: Self.lutSize * Self.lutSize * 4
      )
    }
    return texture
  }

  private static func applyAdjustments(
    _ source: SIMD3<Float>,
    look: NativeGpuCameraLook
  ) -> SIMD3<Float> {
    var color = clamp01(source + SIMD3<Float>(repeating: look.brightness - 1))
    color = clamp01((color - SIMD3<Float>(repeating: midpoint)) * look.contrast + SIMD3<Float>(repeating: midpoint))
    let luminance = color.x * 0.2126 + color.y * 0.7152 + color.z * 0.0722
    color = clamp01(SIMD3<Float>(repeating: luminance) + (color - SIMD3<Float>(repeating: luminance)) * look.saturation)
    return color
  }

  /// Mirrors the existing editor creative kernel's rounded u8 arithmetic.
  private static func applyExactCreative(
    _ source: SIMD3<Float>,
    filterId: String,
    strength: Float
  ) -> SIMD3<Float> {
    let source8 = SIMD3<Int>(
      Int(round(clamp01(source.x) * 255)),
      Int(round(clamp01(source.y) * 255)),
      Int(round(clamp01(source.z) * 255))
    )
    let effected8: SIMD3<Int>
    switch filterId {
    case "grayscale":
      let average = (source8.x + source8.y + source8.z) / 3
      effected8 = SIMD3<Int>(repeating: average)
    case "invert":
      effected8 = SIMD3<Int>(repeating: 255) - source8
    default:
      return source
    }

    let t = max(0, min(1, strength))
    func blended(_ s: Int, _ e: Int) -> Float {
      Float(max(0, min(255, Int(round(Float(s) + Float(e - s) * t))))) / 255
    }
    return SIMD3<Float>(
      blended(source8.x, effected8.x),
      blended(source8.y, effected8.y),
      blended(source8.z, effected8.z)
    )
  }

  private static func sampleAtlas(_ atlas: Data, color: SIMD3<Float>) -> SIMD3<Float> {
    let scaled = clamp01(color) * Float(lutSize - 1)
    let r0 = Int(floor(scaled.x))
    let g0 = Int(floor(scaled.y))
    let b0 = Int(floor(scaled.z))
    let r1 = min(r0 + 1, lutSize - 1)
    let g1 = min(g0 + 1, lutSize - 1)
    let b1 = min(b0 + 1, lutSize - 1)
    let rf = scaled.x - Float(r0)
    let gf = scaled.y - Float(g0)
    let bf = scaled.z - Float(b0)

    func texel(_ red: Int, _ green: Int, _ blue: Int) -> SIMD3<Float> {
      let tileX = blue % tiles
      let tileY = blue / tiles
      let x = tileX * lutSize + red
      let y = tileY * lutSize + green
      let offset = (y * atlasSize + x) * 4
      return atlas.withUnsafeBytes { raw in
        let bytes = raw.bindMemory(to: UInt8.self)
        return SIMD3<Float>(
          Float(bytes[offset]) / 255,
          Float(bytes[offset + 1]) / 255,
          Float(bytes[offset + 2]) / 255
        )
      }
    }

    func slice(_ blue: Int) -> SIMD3<Float> {
      let c00 = texel(r0, g0, blue)
      let c10 = texel(r1, g0, blue)
      let c01 = texel(r0, g1, blue)
      let c11 = texel(r1, g1, blue)
      let top = c00 + (c10 - c00) * rf
      let bottom = c01 + (c11 - c01) * rf
      return top + (bottom - top) * gf
    }

    let low = slice(b0)
    let high = slice(b1)
    return low + (high - low) * bf
  }

  private static func mix(
    _ source: SIMD3<Float>,
    _ effected: SIMD3<Float>,
    amount: Float
  ) -> SIMD3<Float> {
    let t = max(0, min(1, amount))
    return clamp01(source + (effected - source) * t)
  }

  private static func clamp01(_ value: Float) -> Float {
    max(0, min(1, value))
  }

  private static func clamp01(_ value: SIMD3<Float>) -> SIMD3<Float> {
    SIMD3<Float>(clamp01(value.x), clamp01(value.y), clamp01(value.z))
  }

  private static func channelByte(_ value: Float) -> UInt8 {
    UInt8(max(0, min(255, Int(round(clamp01(value) * 255)))))
  }

  private static func error(_ message: String) -> NSError {
    NSError(
      domain: "PixelCraftGpuCameraLook",
      code: 5001,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }
}
